db/version_set.cc

```diff
@@ -2391,6 +2395,31 @@ void VersionStorageInfo::ComputeFilesMarkedForPeriodicCompaction(
   }
 }
 
+void VersionStorageInfo::ComputeFilesMarkedForReadTriggeredCompaction() {
+  files_marked_for_read_triggered_compaction_.clear();
+
+  bool any_files_being_compacted = false;
+  for (int level = 0; level < num_levels(); level++) {
+    for (auto f : files_[level]) {
+      if (f->being_compacted) {
+        any_files_being_compacted = true;
+        continue;
+      }
+      if (ReadThresholdReachedForFile(f)) {
+        files_marked_for_read_triggered_compaction_.emplace_back(level, f);
+      }
+    }
+  }
+  read_triggered_compaction_required_ =
+      !files_marked_for_read_triggered_compaction_.empty();
+  if (!any_files_being_compacted && !read_triggered_compaction_required_) {
```

Comment:
Once compactions are running for all the files that have reached the read threshold, read_triggered_compaction_required_ will be false and files_marked_for_read_triggered_compaction_ will be empty. I didn't want to allow triggering additional compactions from the read path at that point, because they will just end up finding that all the files are already being compacted. As long as compactions are running, they will get re-triggered through the usual path. Only when no more compactions are running we have to re-enable the trigger from the read path.